lib/SILOptimizer/Utils/LexicalDestroyFolding.cpp

```diff
@@ -0,0 +1,772 @@
+    /// The instruction sequence itself.
+    Match match;
+    /// Whether the candidate could indeed be folded as determined by
+    /// isViableInstructionSequence.
+    bool viable;
+    /// The indices of the arguments of the apply to rewrite as determined by
+    /// rewritableArgumentIndicesForApply.
+    SmallVector<int, 2> argumentIndices;
+  };
+
+  enum class MatchViability {
+    // This match can be folded, supposing any can.
+    Viable,
+    // This match cannot be folded, even if others can.
+    Nonviable,
+    // Neither this match nor any other can be folded.
+    Illegal
+  };
+
+  /// Fast check for whether the given instruction is a candidate for folding.
+  ///
+  /// Tries to find patterns like
+  ///
+  ///     apply
+  ///     end_borrow
+  ///     ...           // instructions which CanonicalizeOSSALifetime will not
+  ///                   // hoist destroys over
+  ///     destroy_value %original_borrowee
+  ///
+  /// If None is returned, we can't do the transformation at any end_borrow.
+  /// If a Match WITHOUT an apply is returned, we can't do the transformation
+  /// on the provided instruction but we still might be able to do it on the
+  /// other scope ending instructions.  
+  /// If a Match WITH an apply is returned, we might be able to transform this
+  /// instruction, so more expensive checks are in order.
+  Optional<Match> definesMatchingInstructionSequence(SILInstruction *) const;
+
+  /// Whether the specified instruction is or might be the beginning of a
+  /// sequence of "inconsequential" instructions the last of which destroys
+  /// %original_borrowee.
+  ///
+  /// CanonicalizeOSSALifetime will put destroy_value instructions after every
+  /// final non-consuming use of %original_borrowee.  So it will put a
+  /// destroy_value after an
+  ///     end_borrow %introducer
+  /// if there are no subsequent uses of %original_borrowee.  However, if there
+  /// is already one or more other instructions whose opcodes satisfy
+  /// CanonicalizeOSSALifetime::mustNotHoistDestroyOver just after
+  ///     end_borrow %introducer, 
+  /// it won't hoist the
+  ///     destroy_value %original_borrowee
+  /// over them.
+  ///
+  /// Consequently, it's not good enough just to look at the instruction
+  /// immediately following
+  ///     end_borrow %introducer
+  /// Instead, we need to look through the sequence of them.
+  ///
+  /// To be certain that the sequence is valid, we need to check that none of
+  /// the instructions between the end_borrow and the destroy_value are users of
+  /// %original_borrowee.  That requires having a set of uses to check for
+  /// membership in, though, so that is postponed until
+  /// isViableInstructionSequence.
+  ///
+  /// returns the destroy_value instruction in the sequence that destroys
+  ///         %original_borrowee or nullptr if there isn't one
+  DestroyValueInst *
+  mayBeginOriginalBorroweeDestroyingSequence(SILInstruction *) const;
```

Comment:
nit: How about `findNextBorroweeDestroy`? I mean, there is a comment after all. I fail to parse overly descriptive names.

Reply:
Fixed.  And I agree about names that are too long.  They ought to identify the entity in the relevant context and that's it.